### PR TITLE
[pickers] Fix `DateCalendar` timezone management

### LINKED ISSUE #10804: DateCalendar timezone prop doesn't seem to work (see demo)
### Search keywords

DateCalendar timezone luxon

### Latest version

- [X] I have tested the latest version

### Steps to reproduce

Link to live example: https://codesandbox.io/s/admiring-sutherland-38k683?file=/src/demo.tsx:262-274

Steps:
1. pass a timezone state value as a prop
2. See the calendar render fine: 
image

3. Change the timezone state value
4. See the calendar break (but does not happen for all timezone values): 
image

### Current behavior

_No response_

### Expected behavior

_No response_

### Context

We're trying to create a signup form like Calendly, and this DateCalendar doesn't seem to be handling the switching of time zone correctly.

I'm eager to hear any suggestions. Thanks so much for your help! 

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
 System:
    OS: macOS 13.6
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Browsers:
    Chrome: 118.0.5993.88
    Edge: Not Found
    Safari: 16.6
  npmPackages:
    @emotion/react: ^11.11.1 => 11.11.1 
    @emotion/styled: ^11.11.0 => 11.11.0 
    @mui/base:  5.0.0-beta.17 
    @mui/core-downloads-tracker:  5.14.11 
    @mui/material: ^5.14.1 => 5.14.11 
    @mui/private-theming:  5.14.11 
    @mui/styled-engine:  5.14.11 
    @mui/styled-engine-sc: ^5.12.0 => 5.14.11 
    @mui/system:  5.14.11 
    @mui/types:  7.2.4 
    @mui/utils:  5.14.11 
    @mui/x-date-pickers: ^6.10.1 => 6.15.0 
    @types/react: ^18.2.0 => 18.2.23 
    react: 18.2.0 => 18.2.0 
    react-dom: 18.2.0 => 18.2.0 
    styled-components: ^6.0.5 => 6.0.8 
    typescript: ^4.7.4 => 4.9.5 


```
</details>
Using Chrome Version 118.0.5993.70 (Official Build) (x86_64)

### Order ID or Support key 💳 (optional)

_No response_

### LINKED ISSUE #14453: [pickers] DatePicker displays empty calendar with minDate in future and timezone UTC
### Steps to reproduce

Link to live example: https://codesandbox.io/p/sandbox/zsyzzx

Steps:
1. DatePicker + dayjs. No default value, no initail value.
2. Set `timezone="UTC"` and `minDate` to future (at least tomorrow)
3. Open calendar

### Current behavior

The calendar is empty, literally:
image
The future months as well

### Expected behavior

image

### Context

I'm trying to use DatePicker with `timezone="UTC"` and `minDate`. Sometimes I have `minDate` with value in the future. 

### Your environment

_No response_

**Search keywords**: DatePicker, minDate, timezone

**Search keywords**:

### PR DESCRIPTION
**Pull request by LukasTy**
Fixes #10804.
Fixes #14453.

Closes #12256

The `referenceDate` and `calendarState.currentMonth` were not updating when the `timezone` prop changes.

Changed to:
- update the internal `referenceDate` value when the provided `referenceDate` or timezone prop change
- update internal `calendarState.currentMonth` timezone when `referenceDate` changes

### CODE REVIEW

File: packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          </React.Fragment>
+        );
+      }
+      render(<DateCalendarWithControlledTimezone />);
+
+      expect(
+        screen.getAllByRole('gridcell', {
+          name: (_, element) => element.nodeName === 'BUTTON',
+        }).length,
+      ).to.equal(30);
```
Comment (shaharyar-shamshi) [inline packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx:29]: I think the correct assertion should be no of grid cell of type button before and after the timezone change should be same. (I assume for month of february and month having 31 days it might fail)

and more accurate way of doing this will be before the the timezone change create the map of index of date in gridcell and compare it after the timezone change

File: packages/x-date-pickers/src/DateCalendar/tests/DateCalendar.test.tsx [hunk 2]
```diff
@@ -9,7 +9,9 @@ import { createPickerRenderer, adapterToUse } from 'test/utils/pickers';
 const isJSDOM = /jsdom/.test(window.navigator.userAgent);
 
 describe('<DateCalendar />', () => {
-  const { render, clock } = createPickerRenderer({ clock: 'fake' });
+  const { render, clock } = createPickerRenderer({
```
Comment (shaharyar-shamshi) [inline packages/x-date-pickers/src/DateCalendar/tests/DateCalendar.test.tsx:12]: seems to be unintentional change

File: packages/x-date-pickers/src/DateCalendar/tests/timezone.DateCalendar.test.tsx [hunk 3]
```diff
@@ -29,6 +29,34 @@ describe('<DateCalendar /> - Timezone', () => {
...
+        return (
+          <React.Fragment>
+            <DateCalendar timezone={timezone} />
+            <button onClick={() => setTimezone('America/New_York')}>Switch timezone</button>
+          </React.Fragment>
+        );
+      }
+      render(<DateCalendarWithControlledTimezone />);
+
+      expect(
```
Comment (shaharyar-shamshi) [inline packages/x-date-pickers/src/DateCalendar/tests/timezone.DateCalendar.test.tsx:44]: same as above you can check my PR I have done in same way

**Review by shaharyar-shamshi — COMMENTED** [review]
Added some tentative review. will check further by running it.

**Comment by shaharyar-shamshi** [comment]
I think there is some bug in BasicDateRange select any range (let's assume 12 march to April 21) having UTC as the timezone now change the timezone to America/New_York

only end date is changing not the also when switching to and fro you will notice some of flicker.

@LukasTy

File: packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx (hunk 1 above)
Comment (LukasTy) [reply packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx:29]: Thanks for noting. 🙏 
I've seen your tests and do agree that they are more "complete", but IMHO, I don't see the benefit of having a more complex-to-read test with marginally stricter assertions. The main thing that we need to assert here is that the same amount of days have been rendered after a timezone switch. 🤔 
I do value the benefit of simple tests, which you can glance over and understand in seconds. 🙈 
What do you think?

Comment (shaharyar-shamshi) [reply packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx:29]: I totally agree simple is better. But there might be chance in month of july and August also december 2024 and January 2025 when consecutive months have equal number of dates then this might not cover those edge cases.

Comment (LukasTy) [reply packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx:29]: > But there might be chance in month of july and August also december 2024 and January 2025 when consecutive months have equal number of dates then this might not cover those edge cases.

Why do you think like that? 🤔 
The test is set to use `1` calendar for simplicity and ("current") time is mocked to `2022-06-15`.

Maybe you are talking about a separate issue (https://github.com/mui/mui-x/issues/10584) with AdapterDayjs and DST?
That's a separate topic that I'm looking into. 😉

Comment (shaharyar-shamshi) [reply packages/x-date-pickers-pro/src/DateRangeCalendar/timezone.DateRangeCalendar.test.tsx:29]: No I am not talking about DST. although I missed the date mocking part. I assume it would render the calendar of current. 

With date mocking your test will work. 

Sorry for missing date mocking part.

**Comment by jesse-garrison-panther** [comment]
Hello all, just checking in on this and trying to understand where things are at timeline wise. It looks like this PR fixes an issue I ran [into a while ago](https://github.com/mui/mui-x/issues/12200), so I'm looking forward to this being merged. Is this PR unable to be merged because there is a dependency on a fix for [this other bug](https://github.com/mui/mui-x/issues/10804)?

Thanks for any info you can provide!

**Comment by LukasTy** [comment]
> Hello all, just checking in on this and trying to understand where things are at timeline wise. It looks like this PR fixes an issue I ran [into a while ago](https://github.com/mui/mui-x/issues/12200), so I'm looking forward to this being merged. Is this PR unable to be merged because there is a dependency on a fix for [this other bug](https://github.com/mui/mui-x/issues/10804)?
> 
> Thanks for any info you can provide!

Hey @jesse-garrison-panther, this PR fixes your mentioned problems.
It got delayed because I was looking at fixing a different problem, but it would probably make sense to limit the scope of this PR to only fixing the particular problem with the Calendar months calculation and leave the `AdapterDayjs` issues with DST separate. 🤔

**Comment by jesse-garrison-panther** [comment]
> 
> Hey @jesse-garrison-panther, this PR fixes your mentioned problems. It got delayed because I was looking at fixing a different problem, but it would probably make sense to limit the scope of this PR to only fixing the particular problem with the Calendar months calculation and leave the `AdapterDayjs` issues with DST separate. 🤔

@LukasTy Thanks for the clarification! It would be great if this fix could go out separately from the DST issue you mentioned, although I don't have enough context around that other issue and how it might affect this fix. If they can both be fixed in a similar timeline, that's fine too. I just know personally I've been looking forward to using a new DatePicker component our team built, but the bug I originally filed is preventing us from using it until it's fixed.

Thanks for all the great work you and the team is providing with MUI though!

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx [hunk 4]
```diff
@@ -158,6 +160,17 @@ export const useCalendarState = <TDate extends PickerValidDate>(
     slideDirection: 'left',
   });
 
+  // Ensure that `calendarState.currentMonth` is updated when `referenceDate` (or timezone changes)
+  // https://github.com/mui/mui-x/issues/10804
+  React.useEffect(() => {
+    dispatch({
+      type: 'changeMonth',
+      newMonth: utils.startOfMonth(referenceDate),
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx:168]: Are we sure this change is wanted?
From what I can understand it will reset the displayed month to `referenceDate`, so if the user had moved to another month and then the timezone changes it might be weird.

Tbh this is really an edge case and I could understand that we don't want to optimize for it


____

More problematic, this will break if the component re-renders and `referenceDate` is passed directly is not memoized (which we support for now and we should keep supporting IMHO):

```tsx
import * as React from 'react';
import dayjs from 'dayjs'
import Stack from "@mui/material/Stack";
import Switch from "@mui/material/Switch";
import { DemoContainer } from '@mui/x-date-pickers/internals/demo';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { DateCalendar } from '@mui/x-date-pickers/DateCalendar';

export default function BasicDateCalendar() {
    const [isActive, setIsActive] = React.useState(false)

    return (
        <Stack spacing={2}>
            <Switch value={isActive} onChange={event => setIsActive(event.target.checked)} />
            <LocalizationProvider dateAdapter={AdapterDayjs}>
                <DemoContainer components={['DateCalendar']}>
                    <DateCalendar referenceDate={dayjs('2022-04-17')} />
                </DemoContainer>
            </LocalizationProvider>
        </Stack>
    );
}
```

**Comment by jesse-garrison-panther** [comment]
Hello all, I'm just following up on this issue to see if there is a timeline of when it's expected to be merged? We've built a new date picker component that we're waiting to use, but this bug is currently a blocker. FWIW, we're also part of the Pro Plan. Thanks for any updates you can provide! 🙏

**Comment by LukasTy** [comment]
> Hello all, I'm just following up on this issue to see if there is a timeline of when it's expected to be merged?

@jesse-garrison-panther Thank you for bringing this topic up.
We plan to finish this PR within around a month.
Is that a reasonable timeframe?

**Comment by jesse-garrison-panther** [comment]
> We plan to finish this PR within around a month. Is that a reasonable timeframe?

@LukasTy Thanks for the reply. Yes, I think that timeframe will work.

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx (hunk 4 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx]: @flaviendelangle WDYT about the updated solution: https://github.com/mui/mui-x/pull/12321/commits/f859853266c25cd9dd280ec84d968c6099f0387d?
It would update only the `timezone` and optionally in some cases the `month` (to avoid changing the month when switching timezone to `-the current one`). I.e. `America/New_York` -> `America/Los_Angeles`

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx [hunk 5]
```diff
@@ -54,6 +55,21 @@ export const createCalendarStateReducer =
           isMonthSwitchingAnimating: !reduceAnimations,
         };
 
+      case 'changeMonthTimezone': {
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx:58]: I'm not convinced by the name of this action.
If it runs every time the `referenceDate` updates, could it be named `changeReferenceDate`?
Or if we want to keep the focus on the month, `changeCurrentMonthIfTimezoneChanged`?

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx (hunk 4 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx]: From what I understand, of the two problems I describe above, the 1st one still occurs but the 2nd does not, right?

Comment (LukasTy) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx]: Were you able to replicate the first mentioned problem?
It does not happen to me. 🤔 
This part: https://github.com/mui/mui-x/pull/12321/files#diff-b671464bcd4d0027334e53087da5e27c98342ac1c23b1550a4556879d6f20300R64-R66 is to ensure that the month doesn't shift when `timezone` changes to a minus offset.

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx (hunk 5 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx:58]: > If it runs every time the `referenceDate` updates, could it be named `changeReferenceDate`?

I don't have a strong preference for the action name.
It depends on how we look at the naming:
- Should it refer to what triggered the action?
- Should it refer to what the action will do?

> Or if we want to keep the focus on the month, `changeCurrentMonthIfTimezoneChanged`?

But essentially, it does not change the `currentMonth` most of the time, unless it would cause the month to change because of a timezone change. 🤔

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx (hunk 4 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx]: [Example on master](https://stackblitz.com/edit/react-vsw4h7?file=Demo.tsx)

Change the month to be July using the arrows, then click on "Change timezone".
It should move back to April.

The funny thing is, we already have this behavior on master (I would have thought otherwise), so no regression in your PR :+1:

**Review by flaviendelangle — APPROVED** [review]

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx (hunk 5 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx:58]: If you pass the timezone dirrectly to `changeMonthTimezone` instead of passing the new month, the name and its behavior would then be coherent IMHO

File: packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx (hunk 4 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx]: That's a bit strange behavior that might be worth investigating a bit more in-depth.
I'm a bit puzzled why the current month doesn't change when the current month is changed to `May` (+1). 🤔

Comment (flaviendelangle) [reply packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx]: Not a clue :laughing: 
If you can create an issue and add it to the board so that we don't loose track of it
I don't think it's the top priority right now but I agree it would be interesting to understand what is going on
